Lib/tty.py

```diff
@@ -15,22 +24,81 @@
 OSPEED = 5
 CC = 6
 
-def setraw(fd, when=TCSAFLUSH):
-    """Put terminal into a raw mode."""
-    mode = tcgetattr(fd)
-    mode[IFLAG] = mode[IFLAG] & ~(BRKINT | ICRNL | INPCK | ISTRIP | IXON)
-    mode[OFLAG] = mode[OFLAG] & ~(OPOST)
-    mode[CFLAG] = mode[CFLAG] & ~(CSIZE | PARENB)
-    mode[CFLAG] = mode[CFLAG] | CS8
-    mode[LFLAG] = mode[LFLAG] & ~(ECHO | ICANON | IEXTEN | ISIG)
+try:
+    CHECK_WINSZ = TIOCGWINSZ
+    CHECK_WINSZ = TIOCSWINSZ
+    HAVE_WINSZ = True
+
+    def getwinsz(fd):
+        """Gets window size of tty.
+        If fd is not a descriptor of
+        a tty, then OSError is raised."""
+        s = pack("HHHH", 0, 0, 0, 0)
+        return ioctl(fd, TIOCGWINSZ, s)
+
+    def setwinsz(fd, winsz):
+        """Sets window size of tty.
+        If fd is not a descriptor of
+        a tty, then OSError is raised."""
+        ioctl(fd, TIOCSWINSZ, winsz)
+except NameError:
+    HAVE_WINSZ = False
+    getwinsz = None
+    setwinsz = None
+
+def mode_echo(mode, echo=True):
+    """Set/unset ECHO."""
+    if echo:
+        mode[LFLAG] |= ECHO
+    else:
+        mode[LFLAG] &= ~ECHO
+
+def mode_raw(mode):
+    """Raw mode termios."""
+    mode[IFLAG] &= ~(BRKINT | ICRNL | INPCK | ISTRIP | IXON)
+    mode[OFLAG] &= ~(OPOST)
+    mode[CFLAG] &= ~(CSIZE | PARENB)
+    mode[CFLAG] |= CS8
+    mode[LFLAG] &= ~(ECHO | ICANON | IEXTEN | ISIG)
     mode[CC][VMIN] = 1
     mode[CC][VTIME] = 0
-    tcsetattr(fd, when, mode)
+
+def setraw(fd, when=TCSAFLUSH):
+    """Put terminal into a raw mode.
+    Returns original termios."""
+    mode = tcgetattr(fd)
+    new = list(mode)
+    mode_raw(new)
+    tcsetattr(fd, when, new)
+    return mode
 
 def setcbreak(fd, when=TCSAFLUSH):
     """Put terminal into a cbreak mode."""
     mode = tcgetattr(fd)
-    mode[LFLAG] = mode[LFLAG] & ~(ECHO | ICANON)
-    mode[CC][VMIN] = 1
-    mode[CC][VTIME] = 0
-    tcsetattr(fd, when, mode)
+    new = list(mode)
+    new[LFLAG] &= ~(ECHO | ICANON)
+    new[CC][VMIN] = 1
+    new[CC][VTIME] = 0
+    tcsetattr(fd, when, new)
+    return mode
+
+def login(fd):
```

Comment:
The naming again.
For me, the function does not *login* but *makes the calling process a session leader* as you correctly wrote in docstring.

Maybe `make_session_leader()` or `set_session_leader()` is better?

Reply:
Perfect! I will use one of those, or a variant of one of those. I agree that `tty.login()` was poor choice on my part; I wanted it to sound like `login_tty(3)`.

Reply:
Update: since I had written these a while ago, I had forgotten what this function exactly does. I did some research, and it turns out that this function is actually doing a bit more than making the calling process a session leader.

According to https://pubs.opengroup.org/onlinepubs/9699919799/functions/setsid.html, the `setsid()` call in our function is responsible for making the calling process a session leader; moreover

> Upon return the calling process shall be the session leader of this new session, shall be the process group leader of a new process group, and **shall have no controlling terminal**.

Our function is also making `fd` the stdin/stdout/stderr and **controlling tty** of the calling process, which is why it belongs in the tty module.

Nevertheless, as you said, `login()` sounds like a bad choice; I completely agree with that. Should I change it to `login_tty()`? This function IS actually a clone of `login_tty(3)`.

Additional note: on my Debian system, `man login_tty` says

> These  are BSD functions, present in glibc. They are not standardized in POSIX.

Reply:
Maybe we should add `pty.login_tty()` function instead which calls the native `login_tty(3)`?
It can do some platform-specific things which we don't want to learn and reimplement in Python code.

Reply:
Sir, this sounds like a great idea. I have removed `tty.login_tty()` from this pull request. I have also done restructuring/cleanup. Please check the comments on the diff.

Reply:
Please see #23740 for my implementation of `os.login_tty()`.